### DISCUSSION TITLE
2023/06/10/florr-io-crafting/

### DISCUSSION THREAD

**Discussion by giscus** [argvchs/giscus-comments#30] [Announcements]
# 2023/06/10/florr-io-crafting/

https://argvchs.netlify.app/2023/06/10/florr-io-crafting/

**Comment by Rickyxrc** [comment]
还有一种更简单的方法：

设合卡成功率为 $p$，失败会等概率随机损失 $1$ 到 $5$ 个，即 $2.5$ 个。

那我们定义在合卡概率为 $p$ 下的期望额外损失为 $H(x)$

可以通过定义得到 $H(x) = (1-p)(2.5+H(x))$

解得 $H(x) = \frac{2.5-2.5p}{p}$

首次合成有 $1-p$ 的概率失败，所以最终答案应该是 $5+\frac{2.5(1-p)^2}{p}$

**Reply by argvchs** [reply]
@Rickyxrc ……但是直接推柿子和期望 DP 算出来是不一样的，因为柿子似乎没有考虑到边界的问题

另外还有一种更更简单的方法：

设合卡概率为 $p$，总共合卡 $n$ 次，会消耗 $5n$ 张卡，期望合出卡的数量为 $pn$

但是由于合卡失败返还，期望会合卡失败 $(1 - p)n$ 次，即返还 $2.5(1 - p)n$ 张卡

减去返还后的消耗卡数为 $5n - 2.5(1 - p)n$

则平均消耗 $\dfrac{5n - 2.5(1 - p)n}{pn} = 2.5 + \dfrac{2.5} p$ 张卡期望合出一张

**Reply by Rickyxrc** [reply]
哦哦，是我的锅，问题在首次合成的边界问题。

我改了下上面的函数定义（合成一张卡所需的代价），应该是这样（成功损失 $5$ 张卡，失败损失 $2.5$ 张卡）：

$$\begin{aligned}H(x) &= 5p + (1-p)(2.5+H(x)) \\\\ H(x) &= 2.5 - \frac{2.5}{p}\end{aligned}$$

这样就和您的解法思路一致了。
